src/hp/derp/client.rs

```diff
@@ -0,0 +1,515 @@
+        write_frame_header(&mut *writer, frame_type, 8).await?;
+        writer.write_all(&data).await?;
+        writer.flush().await?;
+        Ok(())
+    }
+
+    /// Sends a packet that tells the server whether this
+    /// client is the user's preferred server. This is only
+    /// used in the server for stats.
+    pub async fn note_preferred(&self, preferred: bool) -> Result<()> {
+        let byte = {
+            if preferred {
+                [0x00]
+            } else {
+                [0x01]
+            }
+        };
+        let mut writer = self.writer.lock().await;
+        write_frame_header(&mut *writer, FRAME_NOTE_PREFERRED, 1).await?;
+        writer.write(&byte).await?;
+        writer.flush().await?;
+        Ok(())
+    }
+
+    /// Sends a request to subscribe to the peer's connection list.
+    /// It's a fatal error if the client wasn't created using [`MeshKey`].
+    pub async fn watch_connection_changes(&self) -> Result<()> {
+        let mut writer = self.writer.lock().await;
+        write_frame_header(&mut *writer, FRAME_WATCH_CONNS, 0).await?;
+        writer.flush().await?;
+        Ok(())
+    }
+
+    /// Asks the server to close the target's TCP connection.
+    /// It's a fatal error if the client wasn't created using [`MeshKey`]
+    pub async fn close_peer(&self, target: PublicKey) -> Result<()> {
+        let mut writer = self.writer.lock().await;
+        write_frame(&mut *writer, FRAME_CLOSE_PEER, target.as_bytes()).await?;
+        Ok(())
+    }
+
+    async fn set_send_rate_limiter(&mut self, sm: ReceivedMessage) {
+        if let ReceivedMessage::ServerInfo {
+            token_bucket_bytes_per_second,
+            ..
+        } = sm
+        {
+            if token_bucket_bytes_per_second == 0 {
+                self.rate_limiter = None;
+            } else {
+                // make a new rate_limiter & add it to the client
+                todo!("the rate limiter stuff is a mess, figure it out");
+            }
+        }
+    }
+
+    async fn local_addr(&self) -> Result<SocketAddr> {
+        {
+            let recv_err = self.recv_err.lock().await;
+            if let Some(e) = &*recv_err {
+                bail!(e.clone());
+            }
+        }
+        Ok(self.conn.local_addr()?)
+    }
+
+    /// Reads a messages from a DERP server.
+    ///
+    /// The returned message may alias memory owned by the [`Client`]; if
+    /// should only be accessed until the next call to [`Client`].
+    ///
+    /// Once [`recv`] returns an error, the [`Client`] is dead forever.
+    pub async fn recv(&self) -> Result<ReceivedMessage> {
+        self.recv_check_error(Duration::from_secs(120)).await
+    }
+
+    async fn recv_check_error(&self, timeout: Duration) -> Result<ReceivedMessage> {
+        {
+            let recv_err = self.recv_err.lock().await;
+            if let Some(err) = &*recv_err {
+                bail!(err.clone());
+            }
```

Comment:
do you really care about the exact error?  We use anyhow errors anyway so i don't think we do.  you can just store an (atomic) bool for the error state for now i think.

Reply:
I... don't know if we care about the exact error. Maybe for now, we go the atomic bool route & just return an anyhow error saying something like "client has encountered a read error and can no longer receive messages"